### ISSUE TITLE
bug yielding Rhat values of "nan"

### ISSUE THREAD

**Issue by jbloom** [stan-dev/pystan#133]
I am using PyStan to do MCMC over some variables in a simplex (Dirichlet prior). I check for convergence by looking at the Rhat value for each variable in the simplex. Occasionally, a few of the variables in the simplex have an Rhat of "nan." Even when I repeatedly re-run the MCMC with more steps, these variables keep having an Rhat of "nan." This is despite the fact that other variables in the simplex will have Rhat values of very close to one (<1.02) and the effective sample size (n_eff) will be >1000 even for the variables returning "nan" for Rhat.

I am speculating that there is some bug in the Rhat calculation that ends up returning "nan" for pathological values of certain variables?

**Comment by jbloom** [comment]
Hi Bob,

I don't think that the problem is just that the variable is unchanged over the MCMC. For instance, at the bottom of this e-mail I have printed out the Rhat, n_eff, 25%, and 75% interval values for the elements of the simplex (Dirichlet prior). You can see that about half the variables in the simplex have a Rhat of nan. However, the n_eff is still quite large and the 25% and 75% interval values are different, so the simplex element is sampling distinct values during the MCMC. 

In general, simplex elements with small numerical values tend to be more likely to produce nan, although there is not a complete correlation.

Note that this is a fairly rare issue. I am running large numbers of MCMC chains with simplexes and monitoring convergence by Rhat, and only rarely does this problem arise. However, it definitely causes a serious problem when trying to write programs that are automatically using Rhat to check for convergence.

Here are the Rhat, n_eff, 25%, and 75% values for the elements of the simplex pir taken from the fitsummary produced by pystan:

For variable pir[0], values are:
	Rhat = 1.00209
	n_eff = 1123
	25% = 0.29137
	75% = 0.329987
For variable pir[1], values are:
	Rhat = nan
	n_eff = 4701
	25% = 4.50462e-05
	75% = 0.00021267
For variable pir[2], values are:
	Rhat = nan
	n_eff = 3336
	25% = 0.000134368
	75% = 0.000385033
For variable pir[3], values are:
	Rhat = 1.00483
	n_eff = 1991
	25% = 0.00107831
	75% = 0.00292147
For variable pir[4], values are:
	Rhat = nan
	n_eff = 2997
	25% = 0.00027064
	75% = 0.000641184
For variable pir[5], values are:
	Rhat = nan
	n_eff = 2185
	25% = 0.00190983
	75% = 0.00260189
For variable pir[6], values are:
	Rhat = 1.00181
	n_eff = 1489
	25% = 0.0850974
	75% = 0.102318
For variable pir[7], values are:
	Rhat = nan
	n_eff = 1767
	25% = 0.00261818
	75% = 0.00380687
For variable pir[8], values are:
	Rhat = nan
	n_eff = 2677
	25% = 0.000831615
	75% = 0.00148543
For variable pir[9], values are:
	Rhat = 1.0053
	n_eff = 1153
	25% = 0.041626
	75% = 0.0525248
For variable pir[10], values are:
	Rhat = nan
	n_eff = 4597
	25% = 6.64422e-05
	75% = 0.000325555
For variable pir[11], values are:
	Rhat = nan
	n_eff = 4269
	25% = 4.96349e-05
	75% = 0.000225117
For variable pir[12], values are:
	Rhat = nan
	n_eff = 2745
	25% = 0.000304088
	75% = 0.00055584
For variable pir[13], values are:
	Rhat = 1.00109
	n_eff = 1694
	25% = 0.010415
	75% = 0.0138256
For variable pir[14], values are:
	Rhat = nan
	n_eff = 2724
	25% = 0.000183746
	75% = 0.0003497
For variable pir[15], values are:
	Rhat = 1.00516
	n_eff = 883
	25% = 0.476673
	75% = 0.518131
For variable pir[16], values are:
	Rhat = nan
	n_eff = 3651
	25% = 5.97385e-05
	75% = 0.000220426
For variable pir[17], values are:
	Rhat = nan
	n_eff = 1689
	25% = 0.00608354
	75% = 0.00662153
For variable pir[18], values are:
	Rhat = 1.03434
	n_eff = 175
	25% = 0.000866166
	75% = 0.00329241
For variable pir[19], values are:
	Rhat = 1.00168
	n_eff = 1859
	25% = 0.0113591
	75% = 0.0151361
For variable pir[20], values are:
	Rhat = nan
	n_eff = 2966
	25% = 0.000309904
	75% = 0.000646834

**Comment by bob-carpenter** [comment]
That definitely looks like a bug.

- Bob

**Comment by ariddell** [comment]
Yes. I want to fix this.

I'm tempted to write a "RFC" for on common Rhat implementation. PyStan tried using Stan's Rhat implementation but it was far too slow (it could not operate on the chains in situ in Python). The current implementation is a copy of RStan's implementation but something is clearly going wrong.

I'd love a C++ rhat function (without Eigen) that both RStan and PyStan could call. The trick is that it has to accept `num_chain` arrays of doubles where the memory is not contiguous (otherwise we could use Eigen).

**Comment by ahartikainen** [comment]
I guess these are the lines (206, 214) in _chains.pyx (functin `split_potential_scale_reduction`) are reason for the behaviour.

if items in `split_chain` (line 206) have small variance Then `var_within` (line 209) has small mean. And it is this mean that is compared against `pystan.constants.EPSILON` that is currently `1e-6`.
  
    # line 206
        split_chain_var.push_back(stan_variance(split_chain))

    cdef double var_between = n_samples / 2 * stan_variance(split_chain_mean)
    # line 209
    cdef double var_within = stan_mean(split_chain_var)

    if var_within < pystan.constants.EPSILON:
        srhat = float('nan')
    else:
        srhat = sqrt((var_between/var_within + n_samples / 2 - 1)/ (n_samples / 2))

Instead if `var_within < pystan.constants.EPSILON` should it compare `var_between/var_within` or  `var_between/stan_mean(split_chain_mean)` some specified limit.

**Comment by jjramsey** [comment]
I noticed that in the function split_potential_scale_reduction(), cython.cdivision is set to True, so there should be no divide by zero tests. If that's the case, then the "if var_within < pystan.constants.EPSILON" condition is probably superfluous and possibly buggy. If there really is a division by zero, then srhat should be nan or inf regardless. However, if var_within is small but valid, then the condition "var_within < pystan.constants.EPSILON" will lead srhat to have a spurious nan value. Also, if both var_within and var_between are zero, then perhaps rhat should be one, since the limit of rhat as var_within and var_between approach zero is one.

**Comment by ariddell** [comment]
Obsoleted by new code.

**Comment by jjramsey** [comment]
I've looked at the source for the development version of PyStan, and it doesn't look like the issue has been obsoleted at all. The function `split_potential_scale_reduction` still compares `var_within` with `pystan.constants.EPSILON`, and that function is still called when PyStan calculates a summary.

Furthermore, I just got bit by the needless comparison to `pystan.constants.EPSILON`, and I would now say that it is not just "possibly buggy" as I said before, but *definitely* buggy. It needs to go.
